server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -175,18 +317,21 @@ private boolean enqueueMergeTaskExecution() {
                         // is also drained, so any queued merge tasks are also forgotten.
                         break;
                     }
-                    // let the task's scheduler decide if it can actually run the merge task now
-                    ThreadPoolMergeScheduler.Schedule schedule = smallestMergeTask.schedule();
-                    if (schedule == RUN) {
-                        runMergeTask(smallestMergeTask);
-                        break;
-                    } else if (schedule == ABORT) {
-                        abortMergeTask(smallestMergeTask);
-                        break;
-                    } else {
-                        assert schedule == BACKLOG;
-                        // the merge task is backlogged by the merge scheduler, try to get the next smallest one
-                        // it's then the duty of the said merge scheduler to re-enqueue the backlogged merge task when it can be run
+                    try (var budgetRelease = budgetTracker.checkBudget(smallestMergeTask)) {
+                        ThreadPoolMergeScheduler.Schedule schedule = smallestMergeTask.schedule();
+                        if (schedule == RUN) {
+                            runMergeTask(smallestMergeTask);
+                            break;
+                        } else if (schedule == ABORT) {
+                            abortMergeTask(smallestMergeTask);
+                            break;
+                        } else {
+                            assert schedule == BACKLOG;
+                            // the merge task is backlogged by the merge scheduler, try to get the next smallest one
+                            // it's then the duty of the said merge scheduler to re-enqueue the backlogged merge task when it can be run
+                        }
+                    } catch (BudgetTracker.BudgetOverflowException e) {
+                        // continue to poll for new merge tasks, over-budget merge task will be reenqueued
```

Comment:
over-budget merge tasks have been "backlogged" until some more budget will become available, at which point **all** will be re-enqueued.